test/python/circuit/library/test_adders.py

```diff
@@ -0,0 +1,107 @@
+                                adder: QuantumCircuit,
+                                inplace: bool,
+                                modular: bool):
+        """Assert that adder correctly implements the summation.
+
+        Args:
+            num_state_qubits: The number of bits in the numbers that are added.
+            adder: The circuit performing the addition of two numbers with ``num_state_qubits``
+                bits.
+            inplace: If True, compare against an inplace addition where the result is written into
+                the second register plus carry qubit. If False, assume that the result is written
+                into a third register of appropriate size.
+            modular: If True, omit the carry qubit to obtain an addition modulo
+                ``2^num_state_qubits``.
+        """
+        circuit = QuantumCircuit(*adder.qregs)
+
+        # create equal superposition
+        circuit.h(range(2 * num_state_qubits))
+
+        # apply adder circuit
+        circuit.compose(adder, inplace=True)
+
+        # obtain the statevector and the probabilities, we don't trace out the ancilla qubits
+        # as we verify that all ancilla qubits have been uncomputed to state 0 again
+        statevector = Statevector(circuit)
+        probabilities = np.abs(statevector) ** 2
+        pad = '0' * circuit.num_ancillas  # state of the ancillas
+
+        # compute the expected results
+        expectations = np.zeros_like(probabilities)
+        num_bits_sum = num_state_qubits + 1
+        # iterate over all possible inputs
+        for x in range(2 ** num_state_qubits):
+            for y in range(2 ** num_state_qubits):
+                # compute the sum
+                addition = (x + y) % (2 ** num_state_qubits) if modular else x + y
+                # compute correct index in statevector
+                bin_x = bin(x)[2:].zfill(num_state_qubits)
+                bin_y = bin(y)[2:].zfill(num_state_qubits)
+                bin_res = bin(addition)[2:].zfill(num_bits_sum)
+                bin_index = pad + bin_res + bin_x if inplace else pad + bin_res + bin_y + bin_x
+                index = int(bin_index, 2)
+                expectations[index] += 1 / 2 ** (2 * num_state_qubits)
+        np.testing.assert_array_almost_equal(expectations, probabilities)
+
+    @data(
+        (3, RippleCarryAdder, True),
+        (5, RippleCarryAdder, True),
+        (3, RippleCarryAdder, True, True),
+        (5, RippleCarryAdder, True, True),
+        (3, QFTAdder, True),
+        (5, QFTAdder, True),
+        (3, QFTAdder, True, True),
+        (5, QFTAdder, True, True),
+        (3, ClassicalAdder, True),
+        (5, ClassicalAdder, True),
+    )
+    @unpack
+    def test_summation(self, num_state_qubits, adder, inplace, modular=False):
+        """Test summation for all implemented adders."""
+        adder = adder(num_state_qubits, modular=True) if modular else adder(num_state_qubits)
```

Comment:
I haven't looked into this but can the existing adders be extended to become modular adders? If so can just add the `modular` arg to them now rather than doing this if/else here.

Reply:
Similar question for whether these can become out-of-place adders instead of the current in-place.

Reply:
Out-of-place should be straightforward: we could just copy the content of register `|b>` to another output register with CX gates, and then do inplace addition of `|a>` and that output register. I'll add that 👍🏻

Reply:
About modular addition: by existing adders do you mean the `WeightedAdder` (which we wanted to rename to accumulate or so) or the `ClassicalAdder`? For the latter we should check if we can make that modular, yeah

Reply:
I made the ClassicalAdder also modular, now all `Adder`s support it (not the `WeightedAdder` since it's not the same kind of adder)

Reply:
great 👍